### PR TITLE
withdrawals as part of the coin selection

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#1861 

# Overview

- 12b28f75b69c4bef47e1d1b7f2b765e9d544f800
  :round_pushpin: **take withdrawals into account, one level earlier, during the coin selection**
  Still to be done:

    - Make sure it's correctly done in the largest-first algorithm
    - Add some test scenario that show the influence of the withdrawal

- 547b1301fe7010cc0285a651aa8448cf88b6abc7
  :round_pushpin: **unit-test the newly introduced 'proportionallyTo'**
  Getting this one wrong would be quite bad :s

- 975f72648879a4f9ee513592e322dd10565f5739
  :round_pushpin: **add unit tests showing how withdrawal impacts the random coin selection**

> :bulb: NOTE
>
> I've chosen not to treat the withdrawal as a single _input_, but more as a "money pot" that is proportionally distributed amongst change output, so that it contributes to every output, based on their size. This is to avoid having a small output consuming the entire withdrawal for itself. Note sure if I'll keep the approach in the end, I'll have the night to think about it.

# Comments

- [ ] TODO: take into account the withdrawal when doing largest-first (and testing it)
- [ ] TODO: echo the error message change on "ErrInputsDepleted" to integration tests relying on that message.

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 1]
```diff
@@ -1878,10 +1878,9 @@ instance Buildable e => LiftHandler (ErrCoinSelection e) where
                 ]
         ErrInputsDepleted ->
             apiError err403 InputsDepleted $ mconcat
-                [ "I had to select inputs to construct the "
-                , "requested transaction. Unfortunately, one output of the "
-                , "transaction depleted all available inputs. "
-                , "Try sending a smaller amount."
+                [ "I cannot select enough UTxO from your wallet to construct "
+                , "an adequate transaction. Try sending a smaller amount or "
+                , "increasing the number of available UTxO."
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:1876]: `ErrInputsDepleted` is now used also when trying to select coins on a null UTxO, despite enough balance in the reward account.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs [hunk 2]
```diff
@@ -112,64 +114,89 @@ random
...
     let nOuts = fromIntegral $ NE.length outs
     let maxN = fromIntegral $ maximumNumberOfInputs opt nOuts
+    let totalOut = sum (getCoin . coin <$> outs)
     randomMaybe <- lift $ runMaybeT $
-        foldM makeSelection (maxN, utxo, []) (descending outs)
+        foldM
+            (\acc out ->
+                let
+                    withdraw = totalWithdraw
+                        `proportionallyTo` (getCoin (coin out) % totalOut)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs]: So here, we are guaranteed that the ratio given as second argument is always `<= 1`

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs [hunk 3]
```diff
@@ -112,64 +114,89 @@ random
...
+            (maxN, utxo, [])
+            (descending outs)
     case randomMaybe of
         Just (maxN', utxo', res) -> do
             (_, sel, remUtxo) <- lift $
                 foldM improveTxOut (maxN', mempty, utxo') (reverse res)
-            guard sel $> (sel, remUtxo)
+            -- NOTE re-assigning total withdrawal to cope with potential
+            -- rounding issues.
+            guard sel $> (sel { withdrawal = totalWithdraw }, remUtxo)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs]: `sel` here is constructed by appending multiple selection together from left to right, using the `Monoid` instance on `CoinSelection`. So in the end, the total withdrawal that we get is the sum of all withdrawals that were assigned to each output, but because `proportionallyTo` rounds **down**, the final sum may end up being slightly less than the total (an error of 1 per output at most) yet, we need it to be _precisely equal_ to the reward's balance without what the Ledger will reject the withdrawal. 

Thus, this is reassigned here, and the possible exceeding will be balanced out during fee balancing anyway.

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs [hunk 4]
```diff
@@ -117,12 +123,35 @@ changeBalance = foldl' addCoin 0 . change
 feeBalance :: CoinSelection -> Word64
 feeBalance sel = inputBalance sel - outputBalance sel - changeBalance sel
 
+-- | Total UTxO balance + withdrawal.
+totalBalance :: Quantity "lovelace" Word64 -> [(TxIn, TxOut)] -> Word64
+totalBalance (Quantity withdraw) inps = balance' inps + withdraw
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:125]: :+1:

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs [hunk 5]
```diff
@@ -117,12 +123,35 @@ changeBalance = foldl' addCoin 0 . change
...
+--
+-- >>> 10 `proportionallyTo` 1%1
+-- 10
+--
+-- >>> 10 `proportionallyTo` 1%2
+-- 5
+--
+-- >>> 10 `proportionallyTo` 1%3
+-- 3
+proportionallyTo :: Integral a => a -> Ratio a -> a
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs]: any chance to have unit test for this or something that uses this inside?

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/LargestFirst.hs [hunk 6]
```diff
@@ -59,11 +69,15 @@ largestFirst opt outs utxo = do
             guard s $> (s, UTxO $ Map.fromList utxo')
         Nothing -> do
             let moneyRequested = sum $ (getCoin . coin) <$> (descending outs)
-            let utxoBalance = fromIntegral $ balance utxo
+            let utxoList = Map.toList $ getUTxO utxo
+            let total = totalBalance withdraw utxoList
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/Primitive/CoinSelection/LargestFirst.hs]: :+1:

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs (hunk 5 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs:149]: ok, so it is used with `totalWithdraw` and checked in random...

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs]: real question, maybe trivial... but why do we need to use `proportionallyTo` at all? why this rule to transfer upon withdrawing proportionally to relative weight in total outs?  Why not to transfer randomly (adding up to withdrawal)? Or almost everything to one plus 1 + 1 + ... + 1? Or evenly? Or create one change?

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs]: if we transfer withdraw to one change we would not have this problem;-) But, when used `proportionallyTo` indeed we need to take care of this... So what happens in detail when  `proportionallyTo` rounding does consume not all withdraw? To which changes this additional `1`s go ? To those that correspond to outputs that were affected by this rounding, right?

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs [hunk 7]
```diff
@@ -151,15 +160,106 @@ spec = do
...
                 , utxoInputs = [oneAda, oneAda, oneAda, oneAda]
                 , txOutputs = 2*oneAda :| [oneAda `div` 2]
+                , totalWithdrawal = 0
                 })
 
+        coinSelectionUnitTest random "withdrawal simple"
+            (Right $ CoinSelectionResult
+                { rsInputs = [1]
+                , rsChange = []
+                , rsOutputs = [2]
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:170]: wy it is not `3` here? :thinking:

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs [hunk 8]
```diff
@@ -151,15 +160,106 @@ spec = do
...
+            (Right $ CoinSelectionResult
+                { rsInputs = [1,1]
+                , rsChange = []
+                , rsOutputs = [2,2]
+                })
+            (CoinSelectionFixture
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [1,1]
+                , txOutputs = 2 :| [2]
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:190]: and here `[3,3]` ?

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs [hunk 9]
```diff
@@ -151,15 +160,106 @@ spec = do
                 , validateSelection = noValidation
                 , utxoInputs = [oneAda, oneAda, oneAda, oneAda]
                 , txOutputs = 2*oneAda :| [oneAda `div` 2]
+                , totalWithdrawal = 0
                 })
 
+        coinSelectionUnitTest random "withdrawal simple"
+            (Right $ CoinSelectionResult
+                { rsInputs = [1]
+                , rsChange = []
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:169]: why it is not `[1]` ?

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs [hunk 10]
```diff
@@ -151,15 +160,106 @@ spec = do
...
+                , validateSelection = noValidation
+                , utxoInputs = [1]
+                , txOutputs = 2 :| []
+                , totalWithdrawal = 1
+                })
+
+        coinSelectionUnitTest random "withdrawal multi-output"
+            (Right $ CoinSelectionResult
+                { rsInputs = [1,1]
+                , rsChange = []
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:183]: why it is not `[1,1]` ?

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs (hunk 9 above)
Comment (paweljakubas) [reply lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:169]: because of the `proportionallyTo` rounding?

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs [hunk 11]
```diff
@@ -151,15 +160,106 @@ spec = do
...
+                })
+            (CoinSelectionFixture
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [1]
+                , txOutputs = 10 :| [10]
+                , totalWithdrawal = 20
+                })
+
+        coinSelectionUnitTest random "withdrawal requires at least one input"
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:253]: this one I understand :+1:

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs (hunk 9 above)
Comment (paweljakubas) [reply lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:169]: ah, ok. we have input 1 plus withdraw 1 to cover 2

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs (hunk 10 above)
Comment (paweljakubas) [reply lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:183]: inps  + withdrawals : `[1,1,2]` to  cover `[2,2]` gives `[]` change

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs [hunk 12]
```diff
@@ -151,15 +160,106 @@ spec = do
...
+                })
+
+        coinSelectionUnitTest random "withdrawal not even"
+            -- 10 Ada available as withdrawal
+            --
+            -- - 7 Ada goes to the first output (10/14 * 10 ~= 7)
+            -- - 2 Ada goes to the other output ( 4/14 * 10 ~= 2)
+            (Right $ CoinSelectionResult
+                { rsInputs = [5,5]
+                , rsChange = [2,3]
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs]: and here we see that in spite of missing `1` by proportionallyTo the money is counting right

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs [hunk 13]
```diff
@@ -151,15 +160,106 @@ spec = do
...
+            (Right $ CoinSelectionResult
+                { rsInputs = [1]
+                , rsChange = [1]
+                , rsOutputs = [10, 10]
+                })
+            (CoinSelectionFixture
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [1]
+                , txOutputs = 10 :| [10]
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs]: in old days it would trigger not enough fragmentation if the number of inputs does not match number of outputs. now I see withdrawals help in this sense :+1:  wonder if there are some integration tests checking this edge conditions that needs to be updated

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelectionSpec.hs [hunk 14]
```diff
@@ -99,6 +109,16 @@ spec = do
         prop "All inputs are used" prop_allInputsAreUsed
         prop "All inputs are used per transaction" prop_allInputsAreUsedPerTx
         prop "Addresses are recycled fairly" prop_fairAddressesRecycled
+
+    describe "proportionallyTo" $ do
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelectionSpec.hs]: ah, this one I was asking above - persuasive :100:

File: lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs [hunk 15]
```diff
@@ -707,10 +709,10 @@ genSelectionFor :: NonEmpty TxOut -> Gen CoinSelection
 genSelectionFor outs = do
     let opts = CS.CoinSelectionOptions (const 100) (const $ pure ())
     utxo <- vector (NE.length outs * 3) >>= genUTxO
-    case runIdentity $ runExceptT $ largestFirst opts outs utxo of
+    withdrawal_ <- genWithdrawal
+    case runIdentity $ runExceptT $ largestFirst opts outs (Quantity withdrawal_) utxo of
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:713]: Dusting is going to affect this somehow ? Or we are confident it is going to be fine?

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs (hunk 5 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs]: it is checked below

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs (hunk 13 above)
Comment (paweljakubas) [reply lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs]: what about this example : 
```
            (CoinSelectionFixture
                { maxNumOfInputs = 100
                , validateSelection = noValidation
                , utxoInputs = [1]
                , txOutputs = 1 :| [1,1,1,1,1]
                , totalWithdrawal = 5
                })
```
?

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs (hunk 5 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection.hs]: Indeed!

File: lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs (hunk 3 above)
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Random.hs]: > To which changes this additional 1s go

Most likely, they'll be used to partially cover fee in the fee balancing algorithm here. What will happen in case of rounding issues is that the transaction will be _slightly_ unbalanced (positively) and have a few lovelace extra on the input side assigned to "nothing". Following the fee balancing algorithm, it'll start by consuming that delta, and then, starts to deplete (proportionally again) change outputs. So in the end, the little exceeding isn't lost but goes directly for fee; which is roughly similar to assigning it to any of the change output, since they'll be depleted for fees regardless.

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs (hunk 10 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs:183]: These examples are indeed a bit tricky to follow, I tried to left some note above each test case to make it a bit easier.

File: lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs (hunk 13 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs]: > what about this example

Following the _current_ logic, each output would receive `0` withdrawal bonus (1/6 * 5, rounded down) and therefore, this would fail on the second output because there wouldn't be enough inputs to select. Although indeed, if instead of using the withdrawal proportionally, we were using them fully to cover "just" the right amount, it'd give better result.

File: lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs (hunk 15 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:713]: I am not sure to follow your concern on this one :thinking:  ?

**Comment by KtorZ** [comment]
bors try

**Review by paweljakubas — APPROVED** [review]
LGTM!
